File: sdk/ios/Sources/DuallerSDK/StorageHandler.swift
import Foundation

/// Handles local storage operations
public final class StorageHandler: Handler {
    public let apiName: String = "dualler.setStorageSync"

    private let defaults = UserDefaults.standard

    public func handle(_ message: BridgeMessage) -> String? {
        guard let payload = message.payload else { return "{}" }

        switch payload["_operation"] as? String {
        case "set":
            let key = payload["_key"] as? String ?? ""
            let data = payload["_data"] as? String ?? ""
            self.defaults.set(data, forKey: key)
            return "{}"
        case "get":
            let key = payload["_key"] as? String ?? ""
            let value = self.defaults.string(forKey: key)
            return "{\"data\":\"\(value ?? "null")\"}"
        case "remove":
            let key = payload["_key"] as? String ?? ""
            self.defaults.removeObject(forKey: key)
            return "{}"
        case "clear":
            let suiteNames = self.defaults.dictionaryRepresentation().keys
            for key in suiteNames {
                self.defaults.removeObject(forKey: key)
            }
            return "{}"
        default:
            return "{}"
        }
    }
}

File: sdk/android/src/main/java/com/dualler/sdk/handler/StorageHandler.kt
package com.dualler.sdk.handler

import com.dualler.sdk.protocol.APIContract
import com.dualler.sdk.protocol.BridgeMessage
import com.dualler.sdk.protocol.SandboxPaths
import android.content.Context
import android.content.SharedPreferences

class StorageHandler(private val context: Context) : Handler {
    override val apiName: String = APIContract.SET_STORAGE_SYNC

    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences(SandboxPaths.STORAGE_KEY, Context.MODE_PRIVATE)
    }

    override fun handle(message: BridgeMessage): String? {
        val payload = message.payload as? Map<*, *> ?: return null
        val key = payload["key"] as? String ?: return null
        val data = payload["data"]

        return when (message.type.name) {
            "SET" -> {
                prefs.edit().putString(key, data.toString()).apply()
                "{}"
            }
            "GET" -> {
                val value = prefs.getString(key, null)
                """{"data":$value}"""
            }
            "REMOVE" -> {
                prefs.edit().remove(key).apply()
                "{}"
            }
            "CLEAR" -> {
                prefs.edit().clear().apply()
                "{}"
            }
            else -> "{}"
        }
    }
}
